Allow specifying devices by DT node label in the shell

mbolivar-ampere opened the pull request:
The purposes of this pull request are:

- resolve a longstanding breakage related to devices getting their names from devicetree node names, which are cumbersome and need not be unique (#51449)
- demonstrate that it can work in a couple of shell implementations (gpio_shell and device_service)

We've all been dealing with this problem for years, ever since the label properties were removed throughout the tree. While handling the fallout of upgrading some internal code to a version of zephyr without label properties, it occurred to me that DT node labels are a perfectly good way to refer to devices, if we had a way to capture them.

This PR gets that done by optionally storing DT node labels alongside their related devices. You can use this to reverse search for a device by DT node label in individual shell commands.

I don't have the bandwidth to update all 66 of our shell command sets, but hopefully by doing the macrobatics necessary to build the infrastructure and updating a couple of subsystem shells for demonstration, individual subsystem maintainers can take it from here.

mbolivar-ampere committed:
devicetree: add first-class node label helpers

Add the following new macros:

- DT_FOREACH_NODELABEL
- DT_FOREACH_NODELABEL_VARGS
- DT_INST_FOREACH_NODELABEL
- DT_INST_FOREACH_NODELABEL_VARGS

These are for-each helpers for iterating over the node labels of a
devicetree node. Since node labels are unique in the entire
devicetree, their token representations can be useful as unique IDs in
code as well.

As a first user of these, add:

- DT_NODELABEL_STRING_ARRAY
- DT_INST_NODELABEL_STRING_ARRAY

The motivating use case for these macros is to allow looking up a
struct device by devicetree node label in Zephyr shell utilities.
The work on the shells themselves is deferred to other patches.

To make working with the string array helpers easier, add:

- DT_NUM_NODELABELS
- DT_INST_NUM_NODELABELS

Signed-off-by: Martí Bolívar <mbolivar@amperecomputing.com>

mbolivar-ampere committed:
kernel: add CONFIG_DEVICE_DT_METADATA and helpers

This option allows you to look up a struct device from any of the
node labels that were attached to the devicetree node used to create
the device, etc.

This is helpful because node labels are a much more human-friendly set
of unique identifiers than the node names we are currently relying on
for use with device_get_binding(). Adding this infrastructure in the
device core allows anyone to make use of it without having to
replicate node label storage and search functions in various places in
the tree. The main use case, however, is for looking up devices by
node label in the shell.

Since there is a footprint penalty associated with storing the node
label metadata, leave this option disabled by default.

Signed-off-by: Martí Bolívar <mbolivar@amperecomputing.com>

mbolivar-ampere committed:
gpio: add support for node label based lookup in the shell

Use some new kernel features to make the experience of finding and
dealing with GPIO devices much more ergonomic by allowing the use of
devicetree node labels to identify GPIO devices by default.

Users who wish to avoid the associated footprint penalty can set
CONFIG_DEVICE_DT_METADATA=n by hand, but I think the convenience is
worth the price as a default. If we're running a shell, then we've
already paid a heavy footprint penalty.

Example output for qemu_cortex_m3:

    uart:~$ gpio devices
    Device           Other names
    gpio@40004000    gpio0
    gpio@40005000    gpio1
    gpio@40006000    gpio2
    gpio@40007000    gpio3
    gpio@40024000    gpio4
    gpio@40025000    gpio5
    gpio@40026000    gpio6

Signed-off-by: Martí Bolívar <mbolivar@amperecomputing.com>

pdgendt commented on kernel/CMakeLists.txt:
```diff
@@ -30,6 +30,12 @@ zephyr_syscall_header_ifdef(
   ${ZEPHYR_BASE}/include/zephyr/kernel/mm/demand_paging.h
 )
 
+zephyr_linker_sources_ifdef(
+  CONFIG_DEVICE_DT_METADATA
+  SECTIONS
+  ${ZEPHYR_BASE}/include/zephyr/linker/device-dt-metadata-section.ld
+)
```
I've had issues with using the `SECTIONS` location, as this is added in the `RAMABLE_REGION` (this is an assumption, I'm no linker expert).

I've created PR #71911 to allow putting sections after the `common-rom` sections (where device structs are located)

mbolivar-ampere replied:
Could you explain what the issue is? We're using `SECTIONS` all over the place, so I'm guessing you mean there are certain situations where it's not working as desired, and this is one of them. If so, can you clarify?

pdgendt replied:
So either you will need to add the section to, for example `include/zephyr/linker/common-rom/common-rom-kernel-devices.ld`, or use something like my proposed `ROM_SECTIONS`, I think.

mbolivar-ampere replied:
Cool PR! OK, I'll check twister results for a similar error and see if your #71911 helps in case I  see them too. Thanks for explaining.

mbolivar-ampere replied:
Hmm, interesting. If nobody has a better idea, I guess I can try to rework this to make a pointer to device_dt_metadata from struct device, like we do with other state.

mbolivar-ampere replied:
@pdgendt you know, it was actually cleaner to do it that way and not use SECTIONS at all. Now updated to just have all the DT metadata be reachable from struct device itself, so we can iterate over devices to iterate over their metadata.

mbolivar-ampere committed:
shell: device_service: print DT metadata

Example output on qemu_cortex_m3:

  devices:
  - uart@4000e000 (READY)
    DT node labels: uart2
  - uart@4000d000 (READY)
    DT node labels: uart1
  - uart@4000c000 (READY)
    DT node labels: uart0
  - gpio@40026000 (READY)
    DT node labels: gpio6
  - gpio@40025000 (READY)
    DT node labels: gpio5
  - gpio@40024000 (READY)
    DT node labels: gpio4
  - gpio@40007000 (READY)
    DT node labels: gpio3
  - gpio@40006000 (READY)
    DT node labels: gpio2
  - gpio@40005000 (READY)
    DT node labels: gpio1
  - gpio@40004000 (READY)
    DT node labels: gpio0

Signed-off-by: Martí Bolívar <mbolivar@amperecomputing.com>

dkalowsk reviewed (dismissed):
Giving an approval.  I don't see the CI tests failing, which makes me wonder if the potential issue identified needs a test case as well.  Either way, the changes look good, make sense, pass all current CI runs, and solves an itch I'm sure most developers run into.

Thanks for the changes!

cfriedt reviewed (dismissed):
LGTM

ubieda reviewed (dismissed):
Tried on `native_sim` and I believe works as expected. Here's a snapshot:

``` shell
uart:~$ device list
devices:
- uart (READY)
  DT node labels: uart0
- gpio@800 (READY)
  DT node labels: gpio0
uart:~$ gpio get gpio0 0
0
uart:~$ gpio set gpio0 0 1
uart:~$ 
```

pdgendt commented on include/zephyr/device.h:
```diff
@@ -429,6 +432,9 @@ struct device {
 	};
 #endif
+#if defined(CONFIG_DEVICE_DT_METADATA) || defined(__DOXYGEN__)
+	const struct device_dt_metadata *dt_meta;
```
Nit: missing doxygen for the field itself.

gmarull reviewed (dismissed):
lgtm, and good to see you back here @mbolivar-ampere

gmarull commented:
btw, likely time to consider https://github.com/zephyrproject-rtos/zephyr/issues/51449 fixed, or fixable.

henrikbrixandersen commented on include/zephyr/device.h:
```diff
@@ -887,6 +893,104 @@ __syscall int device_init(const struct device *dev);
+	/* @brief array of node labels as strings, exactly as they
+	 *        appear in the final devicetree
+	 */
+	const char *nodelabels[];
```
Flexible array members are part of C99. Was this tested on multiple toolchains?

mbolivar-ampere replied:
We can assume C99 anywhere in Zephyr, as far as I know.

> Zephyr does not target a specific version of the C standards; however, the Zephyr codebase makes extensive use of the features newly introduced in the 1999 release of the ISO C standard (ISO/IEC 9899:1999, hereinafter referred to as C99) such as those listed below, effectively requiring the use of a compiler toolchain that supports the C99 standard and above

https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/develop/languages/c/index.rst#language-standards

mbolivar-ampere replied:
To answer your question directly, though: no, I only tested with Zephyr SDK.

henrikbrixandersen commented on include/zephyr/device.h:
```diff
@@ -887,6 +893,104 @@ __syscall int device_init(const struct device *dev);
+ * The maximum length is set so that device_get_by_dt_nodelabel() can
+ * be used from userspace.
+ */
+#define Z_DEVICE_MAX_NODELABEL_LEN Z_DEVICE_MAX_NAME_LEN
```
How come a new define is needed for this? Why not just re-use `Z_DEVICE_MAX_NAME_LEN`?

mbolivar-ampere replied:
It's just for clarity, because this is not a device name.

henrikbrixandersen commented:
@mbolivar-ampere Please rebase on `main`.

jameswalmsley commented:
Really like this, can device_get_binding() do both lookups, (the original and the new nodelabel lookup).

That way the callers won't have to call both APIs manually. This solution is really needed for e.g. zephyr micropython port as well as shell.

This would also mean we wouldn't have to update every shell command manually and as its configurable and not default 'y' developers can opt-in to the behaviour.

mrrosen commented:
> Really like this, can device_get_binding() do both lookups, (the original and the new nodelabel lookup).
> 
> That way the callers won't have to call both APIs manually. This solution is really needed for e.g. zephyr micropython port as well as shell.
> 
> This would also mean we wouldn't have to update every shell command manually and as its configurable and not default 'y' developers can opt-in to the behaviour.
> 
>   

While I agree that using `device_get_binding` to search for any name (node name or node label) would be very convenient, I can also understand if there is a desire to keep that API doing what it's always done (ie, finding a node by its device name). However, I think there really should be with this PR a convenience function that does do the names look up then the node label look up (and then maybe the alias or chosen look up since that would also be nice things to include in the metadata from the dt); a universal lookup in some order for most shells to use, something like `device_get_any_binding` or something. Then shells can just do the swap to a single function and `device_get_binding`'s meaning doesn't change.

kartben commented:
Rebased to solve conflicts due to a new command having been introduced in the gpio shell (#72362), as well as the change from (#72202)

Hopefully this can make it in 3.7 :)

kartben commented:
one more rebase post #70111

kartben commented:
rebased again to pull in hotfix https://github.com/zephyrproject-rtos/zephyr/pull/74126


BTW, @aescolar @nashif, since I am not sure @galak will have the cycles to approve, it might make sense to re-assign. This PR has had plenty of +1s already.

marc-hb commented on include/zephyr/device.h:
```diff
@@ -961,21 +1065,24 @@ __syscall int device_init(const struct device *dev);
+		IF_ENABLED(CONFIG_PM_POLICY_DEVICE_CONSTRAINTS,				\
+			(.pm_constraints_size = (constraints_size_),))			\
+		IF_ENABLED(CONFIG_PM_DEVICE, ({ .pm_base = (pm_),})) /**/		\
+		IF_ENABLED(CONFIG_DEVICE_DT_METADATA,					\
```
Interestingly, this never compiled when BOTH `CONFIG_PM_DEVICE` and `CONFIG_DEVICE_DT_METADATA` were true at the same time, because it was missing a comma between the two. Is that an unusual `CONFIG_` combination?

Fixed yesterday (among other CONFIG_DEVICE_DT_METADATA issue(s)) by @gmarull in commit 91e16f16b35c / PR #74970.

In #70403, I'm now adding both `CONFIG_PM_DEVICE` and `CONFIG_DEVICE_DT_METADATA` to `tests/lib/cpp/cxx/prj.conf`, which would have caught this. This was not a C++ specific bug but who cares - the purpose of that test is maximum coverage anyway.
